### PR TITLE
[WIP] Plone 5 compatibilty

### PR DESCRIPTION
**Pull request by tomgross**
Another attempt to make c.solr compatible with Plone 5. Some tests are still failing but I think this is the best what we have at the moment. Any takers to help fixing the tests in Plone 5?

@karalics @do3cc Are you good to delete the plone5 branch? All the relevant parts have been moved and it is kinde of messed up.

### CODE REVIEW

**Comment by karalics** [comment]
It's OK for me. (deleting plone5 branch )

**Comment by tisto** [comment]
@tomgross I would love to see this getting merged. Do you see any chance to rebase? If you don't have time to work on this, maybe someone else can take over. We could also think about organizing a chipin to fund this...

**Comment by tomgross** [comment]
Hi @tisto this is almost done and works for our Plone 5 installation. The biggest issue ar the test failures resulting from the move of the configuration to plone.app.registry. Any help is appreciated here.

**Comment by tisto** [comment]
@tomgross that's good to hear! I briefly looked into this and it seems that the connection settings that are stored as unicode lead to problems with Solr in the test setup. I tried to add a few encode/decode statements but that did not help so far. Strange that this does not seem to happen  in real life though.

**Comment by tisto** [comment]
@tomgross I would like to put some effort into fixing this. Is it ok if I work on your branch or do you prefer if I create my own branch?

**Comment by tomgross** [comment]
Hi @tisto It is perfectly ok to work on my branch :+1:

File: src/collective/solr/manager.py [hunk 1]
```diff
@@ -1,67 +1,23 @@
...
-from zope.component import getUtility
 from zope.component import queryUtility
 from zope.interface import implements
+from plone.api.portal import get_registry_record
 
 logger = getLogger('collective.solr.manager')
 marker = object()
 
 
-class BaseSolrConnectionConfig(object):
```
Comment (tisto) [inline src/collective/solr/manager.py]: @tomgross why did you remove the SolrConnectionConfig? We need this to support multiple Plone instances in one Zope.

File: src/collective/solr/profiles/default/registry.xml [hunk 2]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0"?>
+<registry>
+  <records interface="collective.solr.interfaces.ISolrSchema"
+           prefix="collective.solr">
+    <value key="active">True</value>
```
Comment (tisto) [inline src/collective/solr/profiles/default/registry.xml]: @tomgross Solr needs to be disabled by default. Otherwise installing c.solr on Plone site creation won't work.

File: src/collective/solr/testing.py [hunk 3]
```diff
@@ -194,11 +198,13 @@ def activateAndReindex(portal):
 COLLECTIVE_SOLR_FIXTURE = CollectiveSolrLayer()
 
 COLLECTIVE_SOLR_INTEGRATION_TESTING = IntegrationTesting(
-    bases=(LEGACY_COLLECTIVE_SOLR_FIXTURE,),
+    bases=(PLONE_APP_REGISTRY_FIXTURE,
```
Comment (tisto) [inline src/collective/solr/testing.py]: @tomgross mixing PLONE_APP_REGISTRY_FIXTURE leads to problems since both PLONE_APP_REGISTRY_FIXTURE and COLLECTIVE_SOLR_INTEGRATION_TESTING use PLONE_FIXTURE as a base layer.

File: src/collective/solr/tests/test_query.py [hunk 4]
```diff
@@ -173,7 +170,6 @@ def testSolrSpecifics(self):
 class QueryTests(TestCase):
 
     def setUp(self):
-        provideUtility(SolrConnectionConfig(), ISolrConnectionConfig)
```
Comment (tisto) [inline src/collective/solr/tests/test_query.py]: @tomgross @do3cc by removing SolrConnnectionConfig here and relying on a plone.app.registry lookup for the connection, we break all the unit tests here. We either have to mock SolrConnectionConfig for all those tests, or re-add the SolrConnectionConfig.

File: src/collective/solr/timeout.py [hunk 5]
```diff
@@ -1,39 +0,0 @@
-# -*- coding: utf-8 -*-
```
Comment (tisto) [inline src/collective/solr/timeout.py]: @tomgross @do3cc why did we remove the timeout functionality here? The configlet test fails because we removed the functionality but kept the test.

File: src/collective/solr/utils.py [hunk 6]
```diff
@@ -4,23 +4,30 @@
...
-from zope.component import queryUtility
 
-from collective.solr.interfaces import ISolrConnectionConfig
+from plone import api
+from collective.solr.interfaces import ISolrSchema
+from zope.component import getUtility
+from plone.registry.interfaces import IRegistry
+
+
+def getConfig():
```
Comment (tisto) [inline src/collective/solr/utils.py]: @tomgross @do3cc the getConfig method is called during the unit tests. the lookup always fails because there is no Plone fixture present at all.

File: src/collective/solr/profiles/default/registry.xml (hunk 2 above)
Comment (tomgross) [inline src/collective/solr/profiles/default/registry.xml]: @tisto I'm ok with changing it to False. This was not not purpose.

File: src/collective/solr/manager.py (hunk 1 above)
Comment (tomgross) [inline src/collective/solr/manager.py]: @tisto Configuration moved to the registry. There is no need for a separate way to config solr and the registry is relative to a Plone instance which should allow multiple Plone instances in on Zope. Did I miss something?

File: src/collective/solr/testing.py (hunk 3 above)
Comment (tomgross) [inline src/collective/solr/testing.py]: @tisto With solr config in the registry we need a functional registry for tests. How it get's there I don't care.

File: src/collective/solr/timeout.py (hunk 5 above)
Comment (tomgross) [inline src/collective/solr/timeout.py]: @tisto It's included in Python 2.6+ see https://github.com/collective/collective.solr/commit/6334d9e007bc576691659e7c22199015b634ea7f

File: src/collective/solr/testing.py (hunk 3 above)
Comment (do3cc) [inline src/collective/solr/testing.py]: I am commenting from far away, but try to run the tests with plone.testing 5.0.0 They might produce more testfailures because of what tisto saw.

Comment (tisto) [inline src/collective/solr/testing.py]: @tomgross it seems the PLONE_APP_REGISTRY fixture bleeds into the other tests which leads to a situation where the solr unit tests should fail but they actually pass (because of the test isolation issue). I have a local branch where I fixed the first problem, but then most of the tests fail because the solr unit tests try to lookup a registry that does not exist. The entire test architecture of collective.solr is build around the SolrConnectionConfig. If we remove that we have to refactor a lot of things...I'm not saying it's not worth the effort. It will just be hard to find someone who actually does the work...We also still have the legacy collective solr fixture which needs to be refactored as well.

File: src/collective/solr/manager.py (hunk 1 above)
Comment (tisto) [inline src/collective/solr/manager.py]: @tomgross in theory yes, in practice no. Things are way more complex than that. By removing the ISolrConnectionConfig you also removed the IZCMLSolrConnection config lookup. This removal requires a considerable refactoring both in the code and in the test setup. 

I tried to mock the registry lookup for the unit tests, but this did not work out... :(

Comment (tomgross) [inline src/collective/solr/manager.py]: Do we really need this complexity? I mean it is just a connection between a CMS and an index based on an api (collective.indexing). It shouldn't be that hard. PloneIntranet moved away from collective.solr because of this complexity.

**Comment by tisto** [comment]
@tomgross @do3cc I further investigated the current test failures in this branch. There are two kind of failures:

1) Test isolation issues (most likely from the plone.app.testing refactoring):

There are a lot of test isolation issues here. I started to just comment out tests that bleed into other tests in order to find the real test failures. I'm down to two test failures. Though, even after fixing those, the isolation issues will remain. Some of them might have been triggered by the plone registry refactoring but I think a fair amount have nothing to do with this pr actually. Maybe we should start cleaning up the master branch before moving to this pr...

2) Test failures because unit tests can not look up the registry.

This can be solved by mocking the registry lookup. I did some tests that look promising but I haven't found the right level of mocking yet. This is doable though.

The good thing so far is that expect one problem with the "active" setting, I did not find a problem in "real life".

cc @goschtl @loechel

**Comment by do3cc** [comment]
What do you mean with plone.app.testing refactoring? The changes in plone.testing? They may fail now while they weren't failing earlier, but this is only, because test isolation was broken silently before.

**Comment by tisto** [comment]
Superseded by: https://github.com/collective/collective.solr/pull/126
